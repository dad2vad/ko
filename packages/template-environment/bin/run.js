#!/usr/bin/env node

require("@kosko/template").run(require("../dist/index.node.cjs").default);
