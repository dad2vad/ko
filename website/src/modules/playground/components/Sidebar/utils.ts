export function isValidName(value: string): boolean {
  return /^[\w-_.]+$/.test(value);
}
