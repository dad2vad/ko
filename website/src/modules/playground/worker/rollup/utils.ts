export function getModuleURLForCDN(module: string) {
  return `https://jspm.dev/${module}`;
}
