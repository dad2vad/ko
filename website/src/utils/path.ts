export const sep = "/";

export function isAbsolute(path: string): boolean {
  return path.startsWith(sep);
}

export function isRelative(path: string): boolean {
  return path.startsWith(`.${sep}`) || path.startsWith(`../${sep}`);
}

export function dirname(path: string): string {
  const index = path.lastIndexOf(sep);
  if (index === -1) return ".";
  return path.substring(0, index) || sep;
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf(sep);
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split(sep).filter(Boolean);
  const toParts = to.split(sep).filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join(sep);
}
