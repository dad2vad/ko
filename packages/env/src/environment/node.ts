import { getRequireExtensions } from "@kosko/require";
import { join } from "node:path";
import { formatPath } from "../paths";
import { toArray } from "@kosko/common-utils";
import {
  createEnvironment,
  createLoaderReducers,
  createReducerList
} from "./base";
import { Environment, Loader, ReducerExecutor, ReducerList } from "./types";
import { cwd as getCwd } from "node:process";

export function getNodeExtensions(): string[] {
  return getRequireExtensions().map((ext) => ext.substring(1));
}

/**
 * Options for creating a Node.js environment.
 *
 * @public
 */
export interface NodeEnvironmentOptions {
  /**
   * Current working directory (CWD).
   *
   * @defaultValue `process.cwd()`
   */
  cwd?: string;
}

export interface InternalNodeEnvironmentOptions extends NodeEnvironmentOptions {
  createReducerExecutor(this: void, reducers: ReducerList): ReducerExecutor;
  requireModule(this: void, environment: Environment, id: string): any;
  mergeValues(this: void, data: any[]): any;
}

export function createNodeEnvironment({
  cwd = getCwd(),
  createReducerExecutor,
  requireModule,
  mergeValues
}: InternalNodeEnvironmentOptions): Environment {
  function requireAllEnvs(template: string, component?: string) {
    if (!environment.env) return [];

    const envs = toArray(environment.env);

    return envs.map((env) => {
      const path = formatPath(template, {
        environment: env,
        ...(component && { component })
      });

      return requireModule(environment, join(environment.cwd, path));
    });
  }

  const loader: Loader = {
    global: () => requireAllEnvs(environment.paths.global),
    component: (name) => requireAllEnvs(environment.paths.component, name)
  };
  const reducers = createReducerList(createLoaderReducers(loader, mergeValues));
  const { reduce } = createReducerExecutor(reducers);
  const environment = createEnvironment({
    cwd,
    extensions: getNodeExtensions(),
    setReducers: reducers.setReducers,
    resetReducers: reducers.resetReducers,
    reduce
  });

  return environment;
}
