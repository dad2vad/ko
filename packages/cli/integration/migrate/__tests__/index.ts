import execa from "execa";
import { join } from "node:path";
import { readFile } from "node:fs/promises";
import { runNodeCLI } from "../../utils/run";

const fixtureDir = join(__dirname, "..", "__fixtures__");

let args: string[];
let result: execa.ExecaReturnValue;
let options: execa.Options;

beforeEach(async () => {
  result = await runNodeCLI(args, {
    ...options,
    cwd: fixtureDir
  });
});

describe("when filename is not given", () => {
  beforeAll(() => {
    args = ["migrate"];
    options = { reject: false };
  });

  test("should return status code 1", () => {
    expect(result.exitCode).toEqual(1);
  });

  test("should print the error", () => {
    expect(result.stderr).toMatchSnapshot();
  });
});

describe("when path is -", () => {
  beforeAll(async () => {
    args = ["migrate", "-f", "-"];
    options = {
      input: await readFile(join(fixtureDir, "only-deployment.yaml"))
    };
  });

  test("should return status code 0", () => {
    expect(result.exitCode).toEqual(0);
  });

  test("should print the output", () => {
    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when path is a file", () => {
  beforeAll(() => {
    args = ["migrate", "-f", "only-deployment.yaml"];
    options = {};
  });

  test("should return status code 0", () => {
    expect(result.exitCode).toEqual(0);
  });

  test("should print the output", () => {
    expect(result.stdout).toMatchSnapshot();
  });
});

describe("when path is a directory", () => {
  beforeAll(() => {
    args = ["migrate", "-f", fixtureDir];
    options = {};
  });

  test("should return status code 0", () => {
    expect(result.exitCode).toEqual(0);
  });

  test("should print the output", () => {
    expect(result.stdout).toMatchSnapshot();
  });
});
