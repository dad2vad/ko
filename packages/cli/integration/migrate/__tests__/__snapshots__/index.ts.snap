// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when filename is not given should print the error 1`] = `
"kosko migrate

Migrate YAML into components

Global Options:
      --cwd        Path of working directory             [string] [default: CWD]
      --log-level  Set log level                      [string] [default: "info"]
      --silent     Disable log output                 [boolean] [default: false]
      --help       Show help                                           [boolean]
      --version    Show version number                                 [boolean]

Options:
  -f, --filename  File, directory to migrate                 [string] [required]
      --esm       Generate ECMAScript module (ESM) files               [boolean]

Examples:
  kosko migrate -f path/to/file  Read from the path
  kosko migrate -f -             Read from stdin

Missing required argument: filename"
`;

exports[`when path is - should print the output 1`] = `
""use strict";

const { Deployment } = require("kubernetes-models/apps/v1/Deployment");

const deployment = new Deployment({
  "metadata": {
    "name": "redis-master",
    "labels": {
      "app": "redis"
    }
  },
  "spec": {
    "selector": {
      "matchLabels": {
        "app": "redis",
        "role": "master",
        "tier": "backend"
      }
    },
    "replicas": 1,
    "template": {
      "metadata": {
        "labels": {
          "app": "redis",
          "role": "master",
          "tier": "backend"
        }
      },
      "spec": {
        "containers": [
          {
            "name": "master",
            "image": "k8s.gcr.io/redis:e2e",
            "resources": {
              "requests": {
                "cpu": "100m",
                "memory": "100Mi"
              }
            },
            "ports": [
              {
                "containerPort": 6379
              }
            ]
          }
        ]
      }
    }
  }
});

module.exports = [deployment];"
`;

exports[`when path is a directory should print the output 1`] = `
""use strict";

const { Deployment } = require("kubernetes-models/apps/v1/Deployment");
const { Service } = require("kubernetes-models/v1/Service");

const deployment = new Deployment({
  "metadata": {
    "name": "redis-master",
    "labels": {
      "app": "redis"
    }
  },
  "spec": {
    "selector": {
      "matchLabels": {
        "app": "redis",
        "role": "master",
        "tier": "backend"
      }
    },
    "replicas": 1,
    "template": {
      "metadata": {
        "labels": {
          "app": "redis",
          "role": "master",
          "tier": "backend"
        }
      },
      "spec": {
        "containers": [
          {
            "name": "master",
            "image": "k8s.gcr.io/redis:e2e",
            "resources": {
              "requests": {
                "cpu": "100m",
                "memory": "100Mi"
              }
            },
            "ports": [
              {
                "containerPort": 6379
              }
            ]
          }
        ]
      }
    }
  }
});

const service = new Service({
  "metadata": {
    "name": "redis-master",
    "labels": {
      "app": "redis",
      "role": "master",
      "tier": "backend"
    }
  },
  "spec": {
    "ports": [
      {
        "port": 6379,
        "targetPort": 6379
      }
    ],
    "selector": {
      "app": "redis",
      "role": "master",
      "tier": "backend"
    }
  }
});

const deployment1 = new Deployment({
  "metadata": {
    "name": "redis-master",
    "labels": {
      "app": "redis"
    }
  },
  "spec": {
    "selector": {
      "matchLabels": {
        "app": "redis",
        "role": "master",
        "tier": "backend"
      }
    },
    "replicas": 1,
    "template": {
      "metadata": {
        "labels": {
          "app": "redis",
          "role": "master",
          "tier": "backend"
        }
      },
      "spec": {
        "containers": [
          {
            "name": "master",
            "image": "k8s.gcr.io/redis:e2e",
            "resources": {
              "requests": {
                "cpu": "100m",
                "memory": "100Mi"
              }
            },
            "ports": [
              {
                "containerPort": 6379
              }
            ]
          }
        ]
      }
    }
  }
});

module.exports = [deployment, service, deployment1];"
`;

exports[`when path is a file should print the output 1`] = `
""use strict";

const { Deployment } = require("kubernetes-models/apps/v1/Deployment");

const deployment = new Deployment({
  "metadata": {
    "name": "redis-master",
    "labels": {
      "app": "redis"
    }
  },
  "spec": {
    "selector": {
      "matchLabels": {
        "app": "redis",
        "role": "master",
        "tier": "backend"
      }
    },
    "replicas": 1,
    "template": {
      "metadata": {
        "labels": {
          "app": "redis",
          "role": "master",
          "tier": "backend"
        }
      },
      "spec": {
        "containers": [
          {
            "name": "master",
            "image": "k8s.gcr.io/redis:e2e",
            "resources": {
              "requests": {
                "cpu": "100m",
                "memory": "100Mi"
              }
            },
            "ports": [
              {
                "containerPort": 6379
              }
            ]
          }
        ]
      }
    }
  }
});

module.exports = [deployment];"
`;
